New Page Request throttling (#6272)
TykTechnologies/tyk-docs: tyk-docs/content/api-management/request-throttling.md

---
title: "Request Throttling"
date: 2025-01-10

> tags: [""]

Reviewer: info missing

Author: Added content in the latest commit.

description: ""
keywords: [""]
aliases:
  - /basic-config-and-security/control-limit-traffic/request-throttling
---

## Introduction

> Tyk's Request Throttling feature provides a mechanism to manage traffic spikes by queuing and automatically retrying client requests that exceed [rate limits]({{< ref "" >}}) or [quotas]({{< ref "" >}}), rather than immediately rejecting them. This helps protect upstream services from sudden bursts and improves the resilience of API interactions during temporary congestion.

Reviewer: missing links

> Tyk's Request Throttling feature provides a mechanism to manage traffic spikes by queuing and automatically retrying client requests that exceed [rate limits]({{< ref "" >}}) or [quotas]({{< ref "" >}}), rather than immediately rejecting them. This helps protect upstream services from sudden bursts and improves the resilience of API interactions during temporary congestion.

Reviewer: Are you sure throttling works with quotas? please confirm as I'm not aware of this

<!-- TODO: Add an image. -->

---

> ## Quick Start

Reviewer: "Quick Start" seems a bit out of context as it's a setting

> ## Quick Start

Reviewer: ```suggestion
## Tutorial: Implementing Request Throttling with Security Policies
```

Author: I have changed it back to `quick start`. The header is too long and unnecessary; I want to maintain a consistent format across pages.

Reviewer: ok but imo it's confusing

In this tutorial, we will configure Request Throttling on a Tyk Security Policy to protect a backend service from sudden traffic spikes. We'll start by defining a basic rate limit on a policy, then enable throttling with specific retry settings to handle bursts exceeding that limit, associate a key with the policy, and finally test the behaviour using simulated traffic. This guide primarily uses the Tyk Dashboard for configuration.

### Prerequisites

- **Docker**: We will run the entire Tyk Stack on Docker. For installation, refer to this [guide](https://docs.docker.com/desktop/setup/install/mac-install/).
- **Git**: A CLI tool to work with git repositories. For installation, refer to this [guide](https://git-scm.com/downloads)
- **Dashboard License**: We will configure Streams API using Dashboard. [Contact support](https://tyk.io/contact/) to obtain a license.
- **Curl, seq and xargs**: These tools will be used for testing.

### Instructions

> 1. **Clone Git Repository:**

Reviewer: What???

Author: Move this to the prerequisite section.

>     The [tyk-demo](https://github.com/TykTechnologies/tyk-demo) repository offers a docker-compose environment you can run locally to explore Tyk streams. Open your terminal and clone the git repository using the below command.

Author: It is not related to Tyk Streams; it is a copy/paste error.

   ```

   - `DASHBOARD_LICENCE`: Add your license key. Contact [support](https://tyk.io/contact/) to obtain a license.

3. **Start Tyk Streams**

    Execute the following command:
    ```bash
    ./up.sh
    ```

    This process will take a couple of minutes to complete and will display some credentials upon completion. Copy the Dashboard **username, password, and API key**, and save them for later use.
    ```
            ▾ Tyk Demo Organisation
                  Username : admin-user@example.org
                  Password : 3LEsHO1jv1dt9Xgf
          Dashboard API Key : 5ff97f66188e48646557ba7c25d8c601
    ```

4. **Verify Setup:**

    Open Tyk Dashboard in your browser by visiting [http://localhost:3000](http://localhost:3000) or [http://tyk-dashboard.localhost:3000](http://tyk-dashboard.localhost:3000) and login with the provided **admin** credentials.

5.  **Create an API:**
    -   Log in to your Tyk Dashboard.
    -   Navigate to **API Management > APIs**.
    -   Click **Add New API**. 
    -   Click **Import**. 
    -   Select **Import Type** as **Tyk API**.
    -   Copy the below content in the text box and click **Import API** to create an API. 

>         ```json

Reviewer: ```suggestion
 <details>
        <summary><b>Click to expand and see the API definition</b></summary>

        ```json
```

                }
            ],
            "x-tyk-api-gateway": {
                "info": {
                    "name": "Request Throttling Test",
                    "state": {
                        "active": true
                    }
                },
                "middleware": {
                    "global": {
                        "contextVariables": {
                            "enabled": true
                        },
                        "trafficLogs": {
                            "enabled": true
                        }
                    }
                },
                "server": {
                    "authentication": {
                        "enabled": true,
                        "securitySchemes": {
                            "authToken": {
                                "enabled": true
                            }
                        }
                    },
                    "listenPath": {
                        "strip": true,
                        "value": "/request-throttling-test/"
                    }
                },
                "upstream": {

>                     "rateLimit": {

Reviewer: What is this RL? isn't it global RL? if it's not relevent to the demo, then we shouldn't put it there

Author: Not required, will remove it.

                        "enabled": false,
                        "per": "10s",
                        "rate": 5
                    },
                    "url": "http://httpbin.org/"
                }
            }
        }

>         ```

Reviewer: ```suggestion
        ```

        </details>
```

    1.  Navigate to **API Security > Policies** in the Tyk Dashboard sidebar.
    2.  Click the **Add Policy** button.
    3.  Under the **1. Access Rights** tab:
        *   In the **Add API Access Rule** section, select the `Request Throttling Test` API
    4.  Scroll down to the **Global Limits and Quota** section (still under the **1. Access Rights** tab):
        *   Set the following values for `Rate Limiting`
        *   Enter `5` into the **Requests (or connection attempts)** field.
        *   Enter `10` into the **Per (seconds):** field.
    5.  Select the **2. Configuration** tab.
    6.  In the **Policy Name** field, enter `Request Throttling Policy`.
    7.  From the **Key expire after** dropdown, select `Key never expire`.
    8.  Click the **Create Policy** button.

7.  **Associate an Access Key with the Policy:**

    1.  Navigate to **API Security > Keys** in the Tyk Dashboard sidebar.
    2.  Click the **Add Key** button.
    3.  Under the **1. Access Rights** tab:
        *   In the **Apply Policy** section, select the `Request Throttling Policy` API
    5.  Select the **2. Configuration** tab.
    6.  In the **Alias** field, enter `Request Throttling Key`.

>     7.  From the **Expires** dropdown, select `Do not expire key`.

Reviewer: Not a good practice. Recommend to put there an hour

Author: Changes Done

    8.  Click the **Create Key** button.
    9.  A pop-up window **"Key created successfully"** will appear displaying the key details. **Copy the Key ID (hash)** value shown and save it securely. You will need this key to make API requests in the following steps.
    10. Click **OK** to close the pop-up.

8. **Test Rate Limit**

>     Before enabling throttling, let's observe the standard rate limiting behaviour. We'll send 10 requests in quick succession using `xargs` to simulate a burst that exceeds our configured limit (5 requests per 10 seconds).

Reviewer: What does it mean "bewfore enabling"? what have we done in the above steps? sorry, I'm confused

>     Before enabling throttling, let's observe the standard rate limiting behaviour. We'll send 10 requests in quick succession using `xargs` to simulate a burst that exceeds our configured limit (5 requests per 10 seconds).

Reviewer: "in quick succession" - can we simply say "in parallel" ?

Author: OK.

>     Before enabling throttling, let's observe the standard rate limiting behaviour. We'll send 10 requests in quick succession using `xargs` to simulate a burst that exceeds our configured limit (5 requests per 10 seconds).

Reviewer: ```suggestion
    So far, we've only created a policy for an API definition and created a key that complies with that policy. Before enabling throttling, let's observe the standard rate limiting behaviour. We'll send 10 requests in quick succession using `xargs` to simulate a burst that exceeds our configured limit (5 requests per 10 seconds).
```

Author: Changes incorporated in the latest commit.

    1.  Open your terminal.
    2.  Execute the following command, replacing `<replace-with-key-id>` with the API Key ID you saved earlier:

>         ```bash

Reviewer: Please explain the test to the user:

You are now going to test throttling by running a command that sends multiple requests to your API in parallel. This will help you verify whether your rate-limiting settings are working as expected.

**What this does:**
	•	seq 10 generates a list of 10 numbers.
	•	xargs -n1 -P10 runs 10 parallel commands — one for each number.
	•	The Authorization header contains a key with throttling settings applied.
	•	The response shows just the HTTP status line (e.g. HTTP/1.1 200 OK or HTTP/1.1 429 Too Many Requests).

        seq 10 | xargs -n1 -P10 -I {} bash -c 'curl -s -I -H "Authorization: <replace-with-key-id>" http://tyk-gateway.localhost:8080/request-throttling-test/ | head -n 1'
        ```

>     3.  **Expected Observation:** You should see some requests succeed with `HTTP/1.1 200 OK`, followed quickly by several requests failing with `HTTP/1.1 429 Too Many Requests` as the rate limit is immediately enforced. The order of `200s` vs `429s` might vary depending upon the processing time, but you will see immediate rejections once the limit is hit.

Reviewer: @sharadregoti this is really nice!!! Well done. Has it been tested?

Author: Yes.

    **Sample Output (Illustrative):**

    ```bash
    HTTP/1.1 429 Too Many Requests
    HTTP/1.1 429 Too Many Requests
    HTTP/1.1 429 Too Many Requests
    HTTP/1.1 429 Too Many Requests
    HTTP/1.1 429 Too Many Requests
    HTTP/1.1 200 OK
    HTTP/1.1 200 OK
    HTTP/1.1 200 OK
    HTTP/1.1 200 OK
    HTTP/1.1 200 OK
    ```

> 9.  **Configure Request Throttling by Updating the Security Policy**

Reviewer: What is this step about? can we explain the user ?

Author: Changes Done

    # --- Noticeable pause here ---
    HTTP/1.1 200 OK
    # --- Noticeable pause here ---
    HTTP/1.1 200 OK
    HTTP/1.1 200 OK
    HTTP/1.1 200 OK
    ```
    *(The exact mix of 200s and 429s on the delayed requests depends heavily on timing relative to the 10-second rate limit window reset and the retry attempts).*

This comparison clearly shows how Request Throttling changes the behaviour from immediate rejection to queued retries, smoothing the traffic flow and potentially allowing more requests to succeed during bursts.

---
## Configuration Options

Request Throttling is configured within Tyk [Security Policies]({{< ref "" >}}) or directly on individual [Access Keys]({{< ref "" >}}).

The configuration involves setting two specific fields:

- `throttle_interval`: Defines the wait time (in seconds) between retry attempts for a queued request.
- `throttle_retry_limit`: Sets the maximum number of retry attempts before the request is rejected.

Both fields must be set to a value of 0 or greater to enable throttling. Setting either to `-1` (the default) disables the feature.

You can configure these settings using either the Tyk Dashboard UI or the Tyk Dashboard API.

> {{< tabs_start >}}

Reviewer: I just hit the tab problem again - searched for text and couldn't find it. THe tab is too long. I think it's better if we do it one after the other in this case

Author: The change, one after the other, required unnecessary scrolling. I have added tabs for the security policy and access key level.

{{< tab_start "Dashboard UI" >}}

> The Tyk Dashboard provides a straightforward interface to set throttling parameters on both Policies and Keys.

Reviewer: Have we wrote this explanation in the guide? are we repeating the same?

**For Policies:**

1.  Navigate to **API Security > Policies** in the Tyk Dashboard sidebar.
2.  Click the **Add Policy** button.
3.  Under the **1. Access Rights** tab:
    *   In the **Add API Access Rule** section, select the required API
4.  Scroll down to the **Global Limits and Quota** section (still under the **1. Access Rights** tab):
    *   To enable `Throttling`, `Rate Limiting` should be configured.

>     *   Enter some value into the **Requests (or connection attempts)** field.

Reviewer: ```suggestion
    *   Field **Requests (or connection attempts)**  - Enter the number of requests you want to allow before rate limit is applied.
```

    *   Enter some value into the **Per (seconds):** field.

    *   Now enable `Throttling` by setting the following values in the `Throttling` section:
    *   Uncheck the `Disable Throttling` checkbox.
    *   Enter some value into the **Throttle retries (or connection attempts)** field.
    *   Enter some value into the **Per (seconds):** field.
5.  Select the **2. Configuration** tab.
6.  In the **Policy Name** field, enter a name.
7.  From the **Key expire after** dropdown, select an option.
8.  Click the **Create Policy** button.

**For Access Keys:**

1.  Navigate to **API Security > Keys** in the Tyk Dashboard sidebar.
2.  Click the **Create Key** button.
3.  Under the **1. Access Rights** tab:
    *   Select **Choose API**.
    *   In the **Add API Access Rule** section, select the required API
4.  Scroll down to the **Global Limits and Quota** section (still under the **1. Access Rights** tab):

>     *   To enable `Throttling`, `Rate Limiting` should be configured.

Reviewer: Not a field name, It should be italic

>     *   To enable `Throttling`, `Rate Limiting` should be configured.

Reviewer: ```suggestion
    *   To enable *Throttling*,  we must configure *Rate Limiting* in the policy.
```

>     *   Enter some value into the **Requests (or connection attempts)** field.

Reviewer: - The setting of RL is in the same level as the explanation. It is a bit confusing to me.

>     *   Enter some value into the **Requests (or connection attempts)** field.

Reviewer: ```suggestion
      *   Enter some value into the **Requests (or connection attempts)** field.
```

>     *   Enter some value into the **Per (seconds):** field.

Reviewer: ```suggestion
      *   Enter some value into the **Per (seconds):** field.
```

    *   Now enable `Throttling` by setting the following values in the `Throttling` section:
    *   Uncheck the `Disable Throttling` checkbox.
    *   Enter some value into the **Throttle retries (or connection attempts)** field.
    *   Enter some value into the **Per (seconds):** field.
5.  Select the **2. Configuration** tab.
6.  In the **Alias** field, enter a name.
7.  From the **Expires** dropdown, select an option.
8.  Click the **Create Key** button.

{{< tab_end >}}

{{< tab_start "Tyk Dashboard API" >}}

You can enable and configure Request Throttling by directly manipulating the Policy object or the Access Key using the Tyk Dashboard API.

**Example Configuration (Policy Object):**

Retrieve the policy object using `GET /api/portal/policies/{POLICY_ID}`. Add or modify the `throttle_interval` and `throttle_retry_limit` fields within the policy JSON object. Then, update the policy using `PUT /api/portal/policies/{POLICY_ID}` with the modified object, or create a new one using `POST /api/portal/policies/`.

> ```json

Reviewer: ```suggestion
```yaml
```

Author: We will use JSON with comments as done [here](https://tyk.io/docs/api-management/event-driven-apis/#tyk-gateway)

  "throttle_interval": 1,       // Wait 1 second between retries
  "throttle_retry_limit": 5,    // Attempt a maximum of 5 retries
  // ... other policy fields ...
}
```

**Example Configuration (Access Key):**

Retrieve the key's session object using `GET /api/keys/{KEY_ID}`. Add or modify the `throttle_interval` and `throttle_retry_limit` fields within the session object JSON. Then, update the key using `PUT /api/keys/{KEY_ID}` with the modified session object.

```json
{
  // ... other session object fields ...
  "throttle_interval": 2,       // Wait 2 seconds between retries
  "throttle_retry_limit": 3,    // Attempt a maximum of 3 retries
  // ... other session object fields ...
}
```

**Explanation:**

*   The first example configures a policy. Any key using this policy will inherit the throttling settings: wait 1 second between retries for queued requests, attempting up to 5 times before failing.
*   The second example configures a specific key's session object directly: wait 2 seconds between retries, attempting up to 3 times. Note: Direct key configuration overrides policy settings for that specific key.

{{< tab_end >}}

{{< tabs_end >}}

---
## How It Works

> TODO: This will work after the changes in Tyk streams is merged. For review you can copy the code and it into a mermaid editor.

Reviewer: Shouldn't depend on streams. Can we do the mermaid change in a separate pr please?

Author: The tyk streams PR has been merged with the mermaid changes. I am resolving this ticket.

This retry cycle repeats until either the request can be successfully processed (if capacity becomes available) or a configured maximum number of retries (`throttle_retry_limit`) is reached. Only after exhausting all retries does Tyk return the `429` error to the client.

Think of it like trying to access a service with two restrictions: a limit on how many people can enter per minute (Rate Limit) and a total daily entry pass limit (Quota). If you arrive and *either* the per-minute entry is full *or* you've used your daily pass allowance, standard behaviour is to turn you away immediately. With Throttling enabled, the service instead asks you to wait briefly (the interval) and tries your entry again shortly, checking *both* the per-minute flow and your pass status again, repeating this a few times (the retry limit) before finally turning you away if access is still restricted.

---
## FAQs

### Can I disable Request Throttling?
    

> Yes, you can. If you set `throttle_interval` and `throttle_retry_limit` values to smaller than `0`, the feature will not work. The default value is `-1` and means it is disabled by default.

Reviewer: IMO it should be part of the configuration options

Author: Let's move the content to the  "configuration options" section and then point the FAQ to the docs.

### Prerequisites

Since we are providing a full example, we recommend installing the following, however, if you have a running deployment as a playground, please feel free to jump to [step 6: set up a policy with throttling]({{< ref "#policy-setup" >}}).

- **Docker**: We will run the entire Tyk Stack on Docker. For installation, refer to this [guide](https://docs.docker.com/desktop/setup/install/mac-install/).
- **Git**: A CLI tool to work with git repositories. For installation, refer to this [guide](https://git-scm.com/downloads)
- **Dashboard License**: We will configure Streams API using Dashboard. [Contact support](https://tyk.io/contact/) to obtain a license.
- **Curl, seq and xargs**: These tools will be used for testing.

### Instructions
#### Start up Tyk stack

1. **Clone Git Repository:**

>     The [tyk-demo](https://github.com/TykTechnologies/tyk-demo) repository offers a docker-compose environment you can run locally to explore Tyk streams. Open your terminal and clone the git repository using the below command.

Reviewer: we should use the other repo - tyk-docker pro

Author: I removed this section of the Tyk Stack setup from the quick start and added a link to the tyk-pro-docker-demo in the prerequisites.

    *   To enable *Throttling*,  we must configure *Rate Limiting* in the policy.
        *   Field **Requests (or connection attempts)**  - Enter the number of requests you want to allow before rate limit is applied.
        *   Field **Per (seconds):** - Enter the time window in seconds during which the number of requests specified above is allowed.
    *   Now enable `Throttling` by setting the following values in the `Throttling` section:
        *   Uncheck the `Disable Throttling` checkbox
        *   Field **Throttle retries (or connection attempts)** - Enter the maximum number of times Tyk should attempt to retry a request after it has been queued due to exceeding a rate limit or quota.
        *   Field **Per (seconds):** - Enter the time interval in seconds Tyk should wait between each retry attempt for a queued request.
5.  Select the **2. Configuration** tab
6.  In the **Policy Name** field, enter a name
7.  From the **Key expire after** dropdown, select an option
8.  Click the **Create Policy** button
</details>

{{< tab_end >}}

{{< tab_start "Access Key" >}}

Note: Direct key configuration overrides policy settings only for that specific key.

> 1.  Navigate to **API Security > Keys** in the Tyk Dashboard sidebar

Reviewer: what about you a screenshot? similar to the policy and collapse the written instructions?

Author: Added in my latest commit.
